lib/runtime/src/distributed.rs

```diff
@@ -227,6 +244,76 @@ impl DistributedRuntime {
     pub fn instance_sources(&self) -> Arc<Mutex<HashMap<Endpoint, Weak<InstanceSource>>>> {
         self.instance_sources.clone()
     }
+
+    /// Add a Prometheus metric to a specific hierarchy's registry
+    pub fn add_prometheus_metric(
+        &self,
+        hierarchy: &str,
+        metric_name: &str,
+        prometheus_metric: Box<dyn prometheus::core::Collector>,
+    ) -> anyhow::Result<()> {
+        let mut registries = self.hierarchy_to_metricsregistry.lock().unwrap();
+        let entry = registries.entry(hierarchy.to_string()).or_default();
+
+        // If a metric with this name already exists for the hierarchy, warn and skip registration
+        if entry.has_metric_named(metric_name) {
+            tracing::warn!(
+                hierarchy = ?hierarchy,
+                metric_name = ?metric_name,
+                "Metric already exists in registry; skipping registration"
+            );
+            return Ok(());
+        }
+
+        // Try to register the metric and provide better error information
+        match entry.prometheus_registry.register(prometheus_metric) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                let error_msg = e.to_string();
+                tracing::error!(
+                    hierarchy = ?hierarchy,
+                    error = ?error_msg,
+                    metric_name = ?metric_name,
+                    "Metric registration failed"
+                );
+                Err(e.into())
+            }
+        }
+    }
+
+    /// Add a callback function to metrics registries for the given hierarchies
+    pub fn register_metrics_callback<F>(&self, hierarchies: Vec<String>, callback: F)
+    where
+        F: Fn() -> anyhow::Result<()> + Send + Sync + 'static,
+    {
+        use std::sync::Arc as StdArc;
+        let callback_arc: StdArc<F> = StdArc::new(callback);
+
+        let mut registries = self.hierarchy_to_metricsregistry.lock().unwrap();
+        for hierarchy in hierarchies {
+            let callback_clone = callback_arc.clone();
+            registries
+                .entry(hierarchy)
+                .or_default()
+                .add_callback(move || (callback_clone)());
+        }
+    }
+
+    /// Execute all callbacks for a given hierarchy key and return their results
+    pub fn execute_metrics_callbacks(&self, hierarchy: &str) -> Vec<anyhow::Result<()>> {
+        let registries = self.hierarchy_to_metricsregistry.lock().unwrap();
+        if let Some(entry) = registries.get(hierarchy) {
+            entry.execute_callbacks()
+        } else {
+            Vec::new()
+        }
+    }
+
+    /// Get all registered hierarchy keys. Private because it is only used for testing.
+    fn get_registered_hierarchies(&self) -> Vec<String> {
+        let registries = self.hierarchy_to_metricsregistry.lock().unwrap();
```

Comment:
If you call `get_registered_hierarchies` a lot, a RWLock would be faster. Not sure if that's necessary.

I would expect all the metrics work to happen from a single thread, where the lock wouldn't be necessary.